spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java

```diff
@@ -84,10 +93,31 @@ public SparkDataFile(Types.StructType type, StructType sparkType) {
     sortOrderIdPosition = positions.get("sort_order_id");
   }
 
+  private void wrapPartitionSpec(GenericRowWithSchema specRow) {
+    // We get all the partition fields, but want to project to the current one
+    StructType wrappedPartitionStruct = specRow.schema();
+
+    if (!wrappedPartitionStruct.equals(currentWrappedPartitionStruct)) {
+      this.currentWrappedPartitionStruct = wrappedPartitionStruct;
+
+      // The original IDs are lost in translation, therefore we apply the ones that we know
```

Comment:
I think it may not work in all cases (e.g. rewriting manifests for an older spec if the union type had forced renames).
Let me think a bit.

Reply:
I think the problem may be that this is trying to fix `SparkDataFile`. There isn't enough information here to project, but that's okay.

I think we need to update the place where `SparkDataFile.partition()` is used, in `RewriteManifestsSparkAction.writeManifest`. That uses a `SparkDataFile`, but it has the spec. We just need to pass in the combined partition struct type for `SparkDataFile` and then add the projection before writing into the new manifest file.

Reply:
Yeah, we need something closer to what we have in `SparkPositionDeltaWrite`.

We can pass `Broadcast<Table> table` to `writeManifest`, which would allow us to construct the union partition type as well as have access to the spec we are trying to write to.

I am not yet sure where the projection should be. We could either change the constructor of `SparkDataFile` and init it there.

```
public SparkDataFile(PartitionSpec spec, Types.StructType partitionType, StructType sparkType) {
  ...
  this.partitionWrapper = new SparkStructLike(partitionType);
  this.partitionProjection = StructProjection.create(partitionType, spec.partitionType());
  ...
}
```

Or pass a precomputed `StructProjection partitionProjection`, which we would build in `writeManifest`.

Reply:
Or we could handle the projection before even wrapping `Row` into `SparkDataFile`. We will need to try out these options and see what makes more sense.

Reply:
Thanks @aokolnychyi for chiming in here, much appreciated. I'm just digging into this code for the first time, so bear with me. Wouldn't it make more sense to either pass in all the historical specs (in the case of the metadata column), or just the latest spec when that's needed? It doesn't feel like a scalable approach to pass in a tuple of all the specs, and then only project the ones that we need.

Reply:
@Fokko, I think what gets passed depends on what's already broadcasted and available (which I suspect is why Anton suggested `Broadcast<Table>`). The most direct route would be to send the combined struct type and the spec (since I think that we always write new manifests with just one spec).

I would probably do this in `toManifest` because I think that we want to reuse the projection across rows. Doing it inside `SparkDataFile` would create a new projection each time, which has needed to be fixed in other places. We don't want to do the struct type comparison for each row.